packages/terra-navigation-layout/src/routingUtils.js

```diff
@@ -0,0 +1,105 @@
+ */
+const reduceRouteConfig = (routeConfig, size) => {
+  if (!routeConfig) {
+    return [];
+  }
+
+  return Object.keys(routeConfig).reduce((availableRoutes, routeKey) => {
+    const config = routeConfig[routeKey];
+
+    let componentConfig;
+    if (typeof (config.component) === 'object') {
+      const configForSize = config.component[size];
+
+      if (configForSize) {
+        componentConfig = configForSize;
+      } else if (config.component.default) {
+        // If no component is defined for the current size, check for a default component
+        // and use it, if one is provided.
+        componentConfig = config.component.default;
+      }
+    }
+
+    if (componentConfig && componentConfig.componentClass) {
+      availableRoutes.push({
+        path: config.path,
+        componentClass: componentConfig.componentClass,
+        componentProps: componentConfig.props,
+        parentPaths: [],
+      });
+    }
+
+    return availableRoutes;
+  }, [])
+  .sort((routeA, routeB) => {
+    if (routeA.path < routeB.path) {
+      return 1;
+    } else if (routeA.path > routeB.path) {
+      return -1;
+    }
+
+    return 0;
+  })
+  .map((route, index, array) => {
+    // The root path '/' cannot have any parent routes, so further processing is unnecessary.
+    if (route.path === '/') {
+      return route;
+    }
+
+    // The path being tested is split into segments. Those segments are then rejoined to produce the set of
+    // potential parent paths. Those potential paths are then filtered against the set of actual paths that
+    // were generated so that only matching parent paths are returned.
+
+    const pathSegments = route.path.split('/');
+
+    const potentialParentPaths = [];
+    // The number of necessary iterations is 2 less than the number of total segments.
+    for (let i = 0, length = pathSegments.length; i < length - 2; i += 1) {
+      pathSegments.pop();
+      potentialParentPaths.push(pathSegments.join('/'));
+    }
+
+    // The root path is added manually as it cannot be produced by the above joins.
+    potentialParentPaths.push('/');
+
+    const matchedParentPaths = array.filter(testRoute => (
+      potentialParentPaths.indexOf(testRoute.path) !== -1
+    ))
+    .map(parentRoute => parentRoute.path);
+
+    return Object.assign({}, route, {
+      parentPaths: matchedParentPaths.reverse(),
+    });
+  });
+};
+
+/**
+ * Determines whether the given pathname matches against any of the given routes configurations.
+ * @param {String} pathname The path to be searched for in the routeConfig.
+ * @param {Array} routes The route data to match against.
+ * @param {String} size The breakpoint size used for configuration parsing.
+ */
+const validateMatchExists = (pathname, routes) => {
+  if (!routes || !pathname) {
+    return false;
+  }
+
+  for (let i = 0, length = routes.length; i < length; i += 1) {
```

Comment:
Can we use the forEach

Reply:
If we did, we couldn't abort the loop when we find the first match.